orc/src/main/java/org/apache/iceberg/orc/ORCSchemaUtil.java

```diff
@@ -0,0 +1,412 @@
+        icebergToOrc.putAll(icebergToOrcMapping("value", orcType.getChildren().get(1)));
+        break;
+    }
+
+    if (orcType.getId() > 0) {
+      // Only add to non-root types.
+      icebergToOrc.put(Integer.parseInt(orcType.getAttributeValue(ICEBERG_ID_ATTRIBUTE)),
+          new OrcColumn(name, orcType));
+    }
+
+    return icebergToOrc;
+  }
+
+
+  private static Optional<TypeDescription> getPromotedType(Type icebergType,
+                                                           TypeDescription originalOrcType) {
+    TypeDescription promotedOrcType = null;
+    if (Type.TypeID.LONG.equals(icebergType.typeId()) &&
+        TypeDescription.Category.INT.equals(originalOrcType.getCategory())) {
+      // Promote: int to long
+      promotedOrcType = TypeDescription.createLong();
+    } else if (Type.TypeID.DOUBLE.equals(icebergType.typeId()) &&
+        TypeDescription.Category.FLOAT.equals(originalOrcType.getCategory())) {
+      // Promote: float to double
+      promotedOrcType = TypeDescription.createDouble();
+    } else if (Type.TypeID.DECIMAL.equals(icebergType.typeId()) &&
+        TypeDescription.Category.DECIMAL.equals(originalOrcType.getCategory())) {
+      // Promote: decimal(P, S) to decimal(P', S) if P' > P
+      Types.DecimalType newDecimal = (Types.DecimalType) icebergType;
+      if (newDecimal.scale() == originalOrcType.getScale() &&
+          newDecimal.precision() > originalOrcType.getPrecision()) {
+        promotedOrcType = TypeDescription.createDecimal()
+            .withScale(newDecimal.scale())
+            .withPrecision(newDecimal.precision());
+      }
+    }
+    return Optional.ofNullable(promotedOrcType);
+  }
+
+  private static boolean isSameType(TypeDescription orcType, Type icebergType) {
+    return Objects.equals(TYPE_MAPPING.get(icebergType.typeId()), orcType.getCategory());
+  }
+
+  /**
+   * Convert an ORC schema to an Iceberg schema. This method handles the convertion from the original
+   * Iceberg colum mapping IDs if present in the ORC column attributes, otherwise, ORC column IDs
+   * will be assigned following ORCs pre-order ID assignment.
+   *
+   * @return the Iceberg schema
+   */
+  static Schema toIceberg(TypeDescription orcSchema) {
+    List<TypeDescription> children = orcSchema.getChildren();
+    List<String> childrenNames = orcSchema.getFieldNames();
+    Preconditions.checkState(children.size() == childrenNames.size(),
+        "Error in ORC file, children fields and names do not match.");
+
+    List<Types.NestedField> icebergFields = Lists.newArrayListWithExpectedSize(children.size());
+    AtomicInteger lastColumnId = new AtomicInteger(orcSchema.getMaximumId());
+    for (int i = 0; i < children.size(); i++) {
+      icebergFields.add(convertOrcToIceberg(children.get(i), childrenNames.get(i),
+          lastColumnId::incrementAndGet));
+    }
+
+    return new Schema(icebergFields);
+  }
+
+  private static Integer icebergID(TypeDescription orcType) {
```

Comment:
It is odd to use `Optional.ofNullable` and then convert back using `orElse(null)`. What about just using an `if`?

Reply:
Or return `Optional` since that's how it is used on line 323.